airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -14,41 +27,135 @@ class ConnectorStateManager:
+        # shared_state is populated. Rather than define how to handle shared_state without a clear use case, we're opting to throw an
+        # error instead and if/when we find one, we will then implement processing of the shared_state value.
+        if shared_state:
+            raise ValueError(
+                "Received a GLOBAL AirbyteStateMessages that contain a shared_state. But this library only ever generated per-STREAM "
+                "STATE messages was not generated this connector. This must be an orchestrator or platform error. GLOBAL state messages "
+                "with shared_state will not be processed correctly. "
+            )
+        self.per_stream_states = per_stream_states
 
-    def get_legacy_state(self) -> MutableMapping[str, Any]:
+    def get_stream_state(self, stream_name: str, namespace: Optional[str]) -> Mapping[str, Any]:
         """
-        Returns a deep copy of the current legacy state dictionary made up of the state of all streams for a connector
-        :return: A copy of the legacy state
+        Retrieves the state of a given stream based on its descriptor (name + namespace) including any global shared state.
+        Stream state takes precedence over shared state when there are conflicts
+        :param stream_name: Name of the stream being fetched
+        :param namespace: Namespace of the stream being fetched
+        :return: The combined shared state and per-stream state of a stream
+        """
+        per_stream_state = self.per_stream_states.get(HashableStreamDescriptor(name=stream_name, namespace=namespace))
+        if per_stream_state:
+            return per_stream_state.dict()
+        return {}
+
+    def get_legacy_state(self) -> Mapping[str, Any]:
         """
-        return copy.deepcopy(self.legacy, {})
+        Using the current per-stream state, creates a mapping of all the stream states for the connector being synced
+        :return: A deep copy of the mapping of stream name to stream state value
+        """
+        return {descriptor.name: per_stream.dict() if per_stream else {} for descriptor, per_stream in self.per_stream_states.items()}
+
+    def update_state_for_stream(self, stream_name: str, namespace: Optional[str], value: Mapping[str, Any]):
+        """
+        Overwrites the state blob of a specific stream based on the provided stream name and optional namespace
+        :param stream_name: The name of the stream whose state is being updated
+        :param namespace: The namespace of the stream if it exists
+        :param value: A stream state mapping that is being updated for a stream
+        :return:
+        """
+        stream_descriptor = HashableStreamDescriptor(name=stream_name, namespace=namespace)
+        self.per_stream_states[stream_descriptor] = AirbyteStateBlob.parse_obj(value)
+
+    @classmethod
+    def _extract_from_state_message(
```

Comment:
I think this method deserves a unit test, wdyt?

Reply:
I guess you'll argue that the initialization test of the ConnectorStateManager already covers many cases.

Reply:
yup that was the intent. We test the initialization which is kind of just a proxy for this method so I think that is sufficient. good callout though!